src/business_logic/transaction/fee.rs

```diff
@@ -0,0 +1,133 @@
+
+    let calldata = [
+        general_config.block_info.sequencer_address.0.clone(),
+        Felt::from_bytes_be(&actual_fee.to_be_bytes()),
+        0.into(),
+    ]
+    .to_vec();
+
+    let entry_point_selector = Felt::from_str_radix(
+        "232670485425082704932579856502088130646006032362877466777181098476241604910",
+        10,
+    )
+    .unwrap();
+
+    let fee_transfer_call = ExecutionEntryPoint::new(
+        fee_token_address,
+        calldata,
+        entry_point_selector,
+        tx_context.account_contract_address.clone(),
+        EntryPointType::External,
+        None,
+        None,
+    );
+
+    let mut resources_manager = ExecutionResourcesManager::default();
+    fee_transfer_call
+        .execute(state, &general_config, &mut resources_manager, &tx_context)
+        .map_err(|_| TransactionError::FeeError("Fee transfer failure".to_string()))
+}
+
+// ----------------------------------------------------------------------------------------
+/// Calculates the fee of a transaction given its execution resources.
+/// We add the l1_gas_usage (which may include, for example, the direct cost of L2-to-L1
+/// messages) to the gas consumed by Cairo resource and multiply by the L1 gas price.
+// ----------------------------------------------------------------------------------------
+
+pub(crate) fn calculate_tx_fee(
+    resources: HashMap<String, Felt>,
+    gas_price: u64,
+    general_config: &StarknetGeneralConfig,
+) -> Result<u64, TransactionError> {
+    let gas_usage = resources
+        .get(&"l1_gas_usage".to_string())
+        .ok_or_else(|| TransactionError::FeeError("Invalid fee value".to_string()))?
+        .to_owned();
+
+    let l1_gas_by_cairo_usage = calculate_l1_gas_by_cairo_usage(general_config, resources)?;
+    let total_l1_gas_usage = gas_usage.to_f64().unwrap() + l1_gas_by_cairo_usage;
+
+    Ok(total_l1_gas_usage.ceil() as u64 * gas_price)
+}
+// ----------------------------------------------------------------------------------------
+/// Calculates the L1 gas consumed when submitting the underlying Cairo program to SHARP.
+/// I.e., returns the heaviest Cairo resource weight (in terms of L1 gas), as the size of
+/// a proof is determined similarly - by the (normalized) largest segment.
+// ----------------------------------------------------------------------------------------
+
+pub(crate) fn calculate_l1_gas_by_cairo_usage(
+    general_config: &StarknetGeneralConfig,
+    cairo_resource_usage: HashMap<String, Felt>,
+) -> Result<f64, TransactionError> {
+    let cairo_resource_fee_weights = general_config.cairo_resource_fee_weights.clone();
+    let cairo_resources_names = cairo_resource_usage
+        .clone()
+        .into_keys()
+        .collect::<HashSet<String>>();
+    let resources_weights = cairo_resource_fee_weights
+        .clone()
+        .into_keys()
+        .collect::<HashSet<String>>();
```

Comment:
I don't think you need `.clone()` on those statements. (You may need to change `max_of_keys()` to accept references instead of values.

Reply:
i am using clone() because the method "into_keys()" takes the ownership of the structure. Regardless of that i change `max_of_keys()` to accept references.

Reply:
Try replacing the function with this:
```rs
    // Ensure that every key in `cairo_resource_usage` is present in ``
    if !general_config
        .cairo_resource_fee_weights
        .iter()
        .map(|(k, _)| k.as_str())
        .all(|k| cairo_resource_usage.contains_key(k))
    {
        return Err(TransactionError::ResourcesError);
    }

    // Convert Cairo usage to L1 gas usage.
    Ok(max_of_keys(
        &cairo_resource_usage,
        &general_config.cairo_resource_fee_weights,
    ))
```
Then you can pass the arguments by reference in both this function and `max_of_keys()`.

Reply:
really cool, thanks for the suggestion